### PR TITLE
More interned string

### PR DESCRIPTION
**Pull request by Eh2406**
This is a series of attempts to remove temporary strings. There is no evidence that this will help with anything, but it just seems cleaner.

Probably easier to read one commit at a time.

### CODE REVIEW

File: src/cargo/core/resolver/context.rs [hunk 1]
```diff
@@ -318,8 +318,8 @@ fn build_requirements<'a, 'b: 'a>(
             uses_default_features: true,
             ..
         } => {
-            if s.features().get("default").is_some() {
-                reqs.require_feature("default")?;
+            if s.features().get(&InternedString::new("default")).is_some() {
```
Comment (alexcrichton) [inline src/cargo/core/resolver/context.rs]: If `Borrow` is implemented for `InternedString` would it allow lookups with `&str` here perhaps?

File: src/cargo/sources/registry/mod.rs [hunk 2]
```diff
@@ -235,14 +234,14 @@ enum Field {
 
 #[derive(Deserialize)]
 struct RegistryDependency<'a> {
-    name: Cow<'a, str>,
```
Comment (alexcrichton) [inline src/cargo/sources/registry/mod.rs]: I think for these the `Cow` was due to the derive above to handle escapes in names if they ever creep in

File: src/cargo/core/resolver/context.rs (hunk 1 above)
Comment (Eh2406) [reply src/cargo/core/resolver/context.rs]: `impl Borrow<str> for InternedString` Hmmm. We are currently Hashing InternedString and str the same. So we could. I have a vain hope that we can find some way to have `O(1)` hashing. but we don't now. So we can always remove `Borrow` if we gain faster hashing.

Another peace is to make a singleton for each constant string we interne, it is mostly just "default".

File: src/cargo/sources/registry/mod.rs (hunk 2 above)
Comment (Eh2406) [reply src/cargo/sources/registry/mod.rs]: So do you want me to leave the `cows` in, incase they are useful; or, do you want me to take them out and we can re-add them if needed?

Currently we are scrubbing data by reing the raw version from `RegistryDependency` then cleaning it into a local and using the local instead.

**Comment by Eh2406** [comment]
Ah... like if the name has a " in it. I will put the cows back in the
morning and see if I can add a test to keep this supported.

**Comment by Eh2406** [comment]
Cows are back, and test is added.

File: src/cargo/core/resolver/context.rs (hunk 1 above)
Comment (alexcrichton) [reply src/cargo/core/resolver/context.rs]: Yeah I think for now having `Borrow` may be the best bet in that it'll hopefully give us a nice ergonomic boost in cases like this and also avoid unnecessary temporary allocations in theory

Comment (Eh2406) [reply src/cargo/core/resolver/context.rs]: Added, and it is an ergonomic boost. It should also save one call to `hash` and 2 cash misses. `InternedString::new` only allocates if the value is unseen, so this was not really allocating in the hot path. But optimizers are odd so maybe removing the unused branch will help.

**Comment by Eh2406** [comment]
I was just looking at some profiles from `>servo cargo generate-lockfile -Z no-index-update` and `RegistryPackage::Deserialize` is ~3% So this clean up may have included some important parts of the code!

~90% was checking if files exist, reading files, or asking git to do the same.

**Comment by alexcrichton** [comment]
Nice! How long does that take locally for you as well? 90% checking for files and such is pretty extreme, and it may be the next scaling bottleneck for Cargo...

**Comment by Eh2406** [comment]
I think it is not so long for evry cargo command. I don't wait that long for other projects. I think it is just that servo has a lot of git dependencies specified. Eche one needs to be checked out and parsed and checked for being a workspace, each time it is **updated**. So I don't think that a `build` command would have to do as much work, not that I tested it yet.  I am sure, if everything was that slow we would hear it from the servo team. :-)

```cmd
servo>cargo +nightly generate-lockfile -Z no-index-update
    Updating git repository `https://github.com/servo/osmesa-src`
    Updating git repository `https://github.com/servo/webrender`
    Updating git repository `https://github.com/servo/gaol`
    Updating git repository `https://github.com/servo/devices`
    Updating git repository `https://github.com/energymon/energymon-rust.git`
    Updating git repository `https://github.com/pcwalton/signpost.git`
    Updating git repository `https://github.com/servo/rust-azure`
    Updating git repository `https://github.com/servo/fontsan`
    Updating git repository `https://github.com/energymon/energymon-sys.git`
```

**Comment by alexcrichton** [comment]
Ah ok that makes more sense!

**Pull request merged**